src/sentry/consumer.py

```diff
@@ -8,62 +8,125 @@
+# We need a unique value to indicate when to stop multiprocessing queue
+# an identity on an object() isn't guaranteed to work between parent
+# and child proc
+_STOP_WORKER = '91650ec271ae4b3e8a67cdc909d80f8c'
+
+
+def handle_preprocess(message):
+    data = message['data']
+    event_id = data['event_id']
+    cache_key = message['cache_key']
+    start_time = message['start_time']
+    process_task = (
+        store_tasks.process_event_from_reprocessing
+        if message['from_reprocessing']
+        else store_tasks.process_event
+    )
+
+    store_tasks._do_preprocess_event(cache_key, data, start_time, event_id, process_task)
+
+
+def handle_process(message):
+    data = message['data']
+    event_id = data['event_id']
+    cache_key = message['cache_key']
+    start_time = message['start_time']
+
+    if message['from_reprocessing']:
+        task = store_tasks.process_event_from_reprocessing
+    else:
+        task = store_tasks.process_event
+
+    task.delay(cache_key=cache_key, start_time=start_time, event_id=event_id)
+
+
+def handle_save(message):
+    data = message['data']
+    event_id = data['event_id']
+    cache_key = message['cache_key']
+    start_time = message['start_time']
+    project_id = data['project']
+
+    store_tasks._do_save_event(cache_key, data, start_time, event_id, project_id)
 
-        task.delay(cache_key=cache_key, start_time=start_time, event_id=event_id)
 
-    def handle_save(self, message):
-        data = message['data']
-        event_id = data['event_id']
-        cache_key = message['cache_key']
-        start_time = message['start_time']
-        project_id = data['project']
+dispatch = {}
+for key, handler in (
+    (settings.KAFKA_PREPROCESS, handle_preprocess),
+    (settings.KAFKA_PROCESS, handle_process),
+    (settings.KAFKA_SAVE, handle_save)
+):
+    topic = settings.KAFKA_TOPICS[key]['topic']
+    dispatch[topic] = handler
 
-        store_tasks._do_save_event(cache_key, data, start_time, event_id, project_id)
+
+def handle_task(task):
+    topic, payload = task
+    handler = dispatch[topic]
+    handler(json.loads(payload))
+
+
+def multiprocess_worker(task_queue):
+    # Configure within each Process
+    configured = False
+
+    while True:
+        task = task_queue.get()
+        if task == _STOP_WORKER:
+            task_queue.task_done()
+            return
+
+        # On first task, configure Sentry environment
+        if not configured:
+            from sentry.runner import configure
+            configure()
+            configured = True
+
+        try:
+            handle_task(task)
+        finally:
+            task_queue.task_done()
+
+
+class ConsumerWorker(AbstractBatchWorker):
+    def __init__(self, concurrency=1):
+        from multiprocessing import Process, JoinableQueue as Queue
+
+        self.concurrency = concurrency
+        self.pool = []
+
+        if self.concurrency > 1:
+            self.task_queue = Queue(1000)
```

Comment:
This seems like it might increase the window for duplicate processing during rebalance? Though I'm not sure that really matters in is case, since _hopefully_ these messages are able to be processed quick enough to be handled during the rebalance timeout?